File: Exec Time/exec_time.py
# Execution Time Calculation Program

MAX_TYPES = 10  # Maximum number of instruction types allowed

# Function to calculate execution time
# Formula: ExecutionTime = ClockCycleTime * Σ(InstructionCount_i * CPI_i)
def calculate_exec_time(t_clk, count, cpi, num_types):
    total_cycles = 0.0
    
    # Compute total cycles
    for i in range(num_types):
        total_cycles += count[i] * cpi[i]
    
    # Return execution time
    return t_clk * total_cycles


def main():
    # Get number of instruction types (must be between 1 and MAX_TYPES)
    while True:
        try:
            num_types = int(input(f"Enter the number of instruction types (1-{MAX_TYPES}): "))
            if 1 <= num_types <= MAX_TYPES:
                break
            else:
                print(f"   Invalid. Please enter a number between 1 and {MAX_TYPES}.")
        except ValueError:
            print(f"   Invalid. Please enter a number between 1 and {MAX_TYPES}.")

    # Get clock cycle time (must be > 0)
    while True:
        try:
            t_clk = float(input("Enter the value of clock cycle time (in second): "))
            if t_clk > 0:
                break
            else:
                print("   Invalid. Please enter a positive number.")
        except ValueError:
            print("   Invalid. Please enter a positive number.")

    # Initialize arrays
    count = []
    cpi = []

    # Get counts and CPIs for each instruction type
    for i in range(num_types):
        # count (must be >= 0)
        while True:
            try:
                c = int(input(f"Enter the count of Type {i + 1} instruction: "))
                if c >= 0:
                    count.append(c)
                    break
                else:
                    print("   Invalid. Please enter a non-negative whole number.")
            except ValueError:
                print("   Invalid. Please enter a non-negative whole number.")
        
        # cpi (must be > 0)
        while True:
            try:
                cp = float(input(f"Enter the CPI of Type {i + 1} instruction: "))
                if cp > 0:
                    cpi.append(cp)
                    break
                else:
                    print("   Invalid. Please enter a positive number.")
            except ValueError:
                print("   Invalid. Please enter a positive number.")

    # Calculate execution time using the function
    exec_time = calculate_exec_time(t_clk, count, cpi, num_types)

    # Output result
    print("\n-------------------------------------------")
    print(f"The execution time of this program is {exec_time:.6f} seconds.")
    print("-------------------------------------------")


if __name__ == "__main__":
    main()

File: Exec Time/exec_time.c
#include <stdio.h>

#define MAX_TYPES 10  // Maximum number of instruction types allowed

// Function prototype
double calculate_exec_time(double t_clk, long long count[], double cpi[], int num_types);

int main(void) {
    int num_types;
    double t_clk;
    long long count[MAX_TYPES];
    double cpi[MAX_TYPES];
    double exec_time = 0.0;

    // Get number of instruction types (must be between 1 and MAX_TYPES)
    while (1) {
        printf("Enter the number of instruction types (1-%d): ", MAX_TYPES);
        if (scanf("%d", &num_types) == 1 && num_types >= 1 && num_types <= MAX_TYPES) {
            break;
        } else {
            printf("   Invalid. Please enter a number between 1 and %d.\n", MAX_TYPES);
            while (getchar() != '\n');
        }
    }

    // Get clock cycle time (must be > 0)
    while (1) {
        printf("Enter the value of clock cycle time (in second): ");
        if (scanf("%lf", &t_clk) == 1 && t_clk > 0) {
            break;
        } else {
            printf("   Invalid. Please enter a positive number.\n");
            
            while (getchar() != '\n');
        }
    }

    // Get counts and CPIs for each instruction type
    for (int i = 0; i < num_types; i++) {

        // count (must be >= 0)
        while (1) {
            printf("Enter the count of Type %d instruction: ", i + 1);
            if (scanf("%lld", &count[i]) == 1 && count[i] >= 0) {
                break;
            } else {
                printf("   Invalid. Please enter a non-negative whole number.\n");
                while (getchar() != '\n');
            }
        }

        // cpi (must be > 0)
        while (1) {
            printf("Enter the CPI of Type %d instruction: ", i + 1);
            if (scanf("%lf", &cpi[i]) == 1 && cpi[i] > 0) {
                break;
            } else {
                printf("   Invalid. Please enter a positive number.\n");
                while (getchar() != '\n');
            }
        }
    }

    // Calculate execution time using the function 
    // Formula: ExecutionTime = ClockCycleTime * Σ(InstructionCount_i * CPI_i)
    exec_time = calculate_exec_time(t_clk, count, cpi, num_types);

    // Output result
    printf("\n-------------------------------------------\n");
    printf("The execution time of this program is %.6f seconds.\n", exec_time);
    printf("-------------------------------------------\n");

    return 0;
}

// Function to calculate execution time
double calculate_exec_time(double t_clk, long long count[], double cpi[], int num_types) {
    double total_cycles = 0.0;
    
    // Compute total cycles
    for (int i = 0; i < num_types; i++) {
        total_cycles += count[i] * cpi[i];
    }
    
    // Return execution time
    return t_clk * total_cycles;
}